File: code/jvm/im-domain/src/main/kotlin/im/domain/wrappers/name/NameValidator.kt
package im.domain.wrappers.name

import im.domain.Either
import im.domain.failure
import im.domain.success

class NameValidator(
    private val maxNameLength: Int = MAX_NAME_LENGTH,
    private val minNameLength: Int = MIN_NAME_LENGTH,
) {
    companion object {
        private const val MAX_NAME_LENGTH = 30
        private const val MIN_NAME_LENGTH = 3
    }

    fun validate(value: String): Either<List<NameValidationError>, Unit> {
        val errors = mutableListOf<NameValidationError>()

        if (value.isBlank()) {
            errors.add(NameValidationError.Blank)
        }

        if (value.length !in minNameLength..maxNameLength) {
            errors.add(NameValidationError.InvalidLength(minNameLength, maxNameLength))
        }

        if (errors.isNotEmpty()) {
            return failure(errors)
        }

        return success(Unit)
    }
}

File: code/js/src/Domain/wrappers/name/NameValidator.ts
import { NameValidationError } from './NameValidationError';

export class NameValidator {
    private static readonly MAX_NAME_LENGTH = 30;
    private static readonly MIN_NAME_LENGTH = 3;

    constructor(
        readonly maxNameLength: number = NameValidator.MAX_NAME_LENGTH,
        readonly minNameLength: number = NameValidator.MIN_NAME_LENGTH,
    ) {}

    validate(value: string): NameValidationError[] {
        const errors: NameValidationError[] = [];

        if (value.trim() === '') {
            errors.push(NameValidationError.Blank);
        }

        if (
            value.length < this.minNameLength ||
            value.length > this.maxNameLength
        ) {
            errors.push(
                NameValidationError.InvalidLength(
                    this.minNameLength,
                    this.maxNameLength,
                ),
            );
        }

        return errors.length > 0 ? errors : [];
    }
}
